pkg/compiler/global_test.go

```diff
@@ -80,6 +81,536 @@ func TestUnusedGlobal(t *testing.T) {
+			t.Run("drop value", func(t *testing.T) {
+				src := `package foo
+			var (
+				A = 1
+				B = A + 1 // To check nested ident values.
+				C = 3
+				D = B + f() + C // To check that both idents (before and after the call to f) will be marked as "used".
+			)
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 2
+			}`
+				eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{3}}, // sslot for A
+					opcode.PUSH1, opcode.STSFLD0, // store A
+					opcode.LDSFLD0, opcode.PUSH1, opcode.ADD, opcode.STSFLD1, // store B
+					opcode.PUSH3, opcode.STSFLD2, // store C
+					opcode.LDSFLD1, []interface{}{opcode.CALL, []byte{9}}, opcode.ADD, opcode.LDSFLD2, opcode.ADD, opcode.DROP, opcode.RET, // evaluate D and drop
+					opcode.PUSH1, opcode.RET, // Main
+					opcode.PUSH2, opcode.RET) // f
+			})
+			t.Run("drop value", func(t *testing.T) {
+				src := `package foo
+			var (
+				A = 1
+				C = f() + A		// Unused but contains function call, so value should be evaluated and DROPped.
+				D = C + 1 		// Unused, no code expected.
+			)
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 2
+			}`
+				eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{1}}, // sslot for A
+					opcode.PUSH1, opcode.STSFLD0, // store A
+					[]interface{}{opcode.CALL, []byte{8}}, opcode.LDSFLD0, opcode.ADD, opcode.DROP, opcode.RET, // evaluate C and store
+					opcode.PUSH1, opcode.RET, // Main
+					opcode.PUSH2, opcode.RET) // f
+			})
+		})
+		t.Run("named, with dependency on unused var ident inside function call", func(t *testing.T) {
+			src := `package foo
+			var A = 1
+			var B = f(A)
+			func Main() int {
+				return 1
+			}
+			func f(a int) int {
+				return a
+			}`
+			eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{1}}, // sslot for A
+				opcode.PUSH1, opcode.STSFLD0, // store A
+				opcode.LDSFLD0, []interface{}{opcode.CALL, []byte{6}}, opcode.DROP, opcode.RET, // evaluate B and drop
+				opcode.PUSH1, opcode.RET, // Main
+				[]interface{}{opcode.INITSLOT, []byte{0, 1}}, opcode.LDARG0, opcode.RET) // f
+		})
+		t.Run("named, inside multi-specs and multi-vals var declaration", func(t *testing.T) {
+			src := `package foo
+			var (
+				Unused = 1
+				Unused1, A, Unused2 = 2, 3 + f(), 4
+			)
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 5
+			}`
+			eval(t, src, big.NewInt(1), opcode.PUSH3, []interface{}{opcode.CALL, []byte{7}}, opcode.ADD, opcode.DROP, opcode.RET, // evaluate and drop A
+				opcode.PUSH1, opcode.RET, // Main
+				opcode.PUSH5, opcode.RET) // f
+		})
+		t.Run("unnamed + unused", func(t *testing.T) {
+			src := `package foo
+			var A = 1 // At least one global variable is used, thus, the whole set of variables will be walked
+			var B = 2
+			var _ = B + 1 // This variable is unnamed, thus its children won't be marked as "used" => compilation panic on attempt to walk through B+1 expression.
```

Comment:
Does it really panic now?

Reply:
No, now it works as expected, this comment is a debugging remnant.